src/gpu/nodeKernels.cu

```diff
@@ -229,25 +229,32 @@ __global__ void kProcessBeamSamplesFirstLast(size_t beamCount, int samplesPerBea
 	LIMIT(beamCount);
 
 	const auto beamIdx = tid;
-	int firstIdx = 0;
-	int lastIdx = 0;
+	int firstIdx = -1;
+	int lastIdx = -1;
 	for (int sampleIdx = 0; sampleIdx < samplesPerBeam; ++sampleIdx) {
 		if (beamSamples.isHit[beamIdx * samplesPerBeam + sampleIdx] == 0) {
 			continue;
 		}
-		if (beamSamples.distance[beamIdx * samplesPerBeam + sampleIdx] <
-		    beamSamples.distance[beamIdx * samplesPerBeam + firstIdx]) {
+		auto currentFirstDistance = firstIdx >= 0 ? beamSamples.distance[beamIdx * samplesPerBeam + firstIdx] : FLT_MAX;
+		auto currentLastDistance = lastIdx >= 0 ? beamSamples.distance[beamIdx * samplesPerBeam + lastIdx] : -FLT_MAX;
+		if (beamSamples.distance[beamIdx * samplesPerBeam + sampleIdx] < currentFirstDistance) {
 			firstIdx = sampleIdx;
 		}
-		if (beamSamples.distance[beamIdx * samplesPerBeam + sampleIdx] >
-		    beamSamples.distance[beamIdx * samplesPerBeam + lastIdx]) {
+		if (beamSamples.distance[beamIdx * samplesPerBeam + sampleIdx] > currentLastDistance) {
```

Comment:
Technically this if-statement and the above one exclude themselves - else-if could result is slightly better code, but it would only work assuming that we have current first and last distances already set. This can be done setting firstIdx and lastIdx (lines 232-233) to 0, and skipping sample at index 0 in for-loop in line 234.